File: Python/Trees/IterTrees/In-order.py
from typing import Optional

"""
 * The following impl is also the same time and space complexity as the recursive iteration (O(n) space and O(n) time (prove using divide and conquer strong induction)).
 * Difference from the normal recursive implementation: we have our own cute lil stack.
 *
 * Space is theta(log n) and O(n) no omega.
 * What to prove: T(n) = T(k) + T(n-k-1) + O(1) = O(n) (omega in this case).
 """


class BSTIterator:

    def __init__(self, root: Optional[TreeNode]):
        self.root = root
        self.res = []

        stack = []
        # add the values appropriatly
        curr = self.root
        while curr or stack:
            if curr:
                stack.append(curr)
                curr = curr.left
            else:
                curr = stack.pop()
                self.res.append(curr.val)
                curr = curr.right

    def next(self) -> int:
        return self.res.pop(0)

    def hasNext(self) -> bool:
        if self.res:
            return True
        return False

File: c++/Trees/IterTraversal/in-order.cpp
#include <vector>
#include <stack>
#include <queue>

using namespace std;

/**
 * The following impl is also the same time and space complexity as the recursive iteration (O(n) space and O(n) time (prove using divide and conquer strong induction)).
 * Difference from the normal recursive implementation: we have our own cute lil stack.
 *
 * What to prove: T(n) = T(k) + T(n-k-1) + O(1) = O(n) (omega in this case).
 */

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class BSTIterator
{
public:
    BSTIterator(TreeNode *root)
    {
        TreeNode *curr = root;

        while (curr != NULL || stack.size() > 0)
        {
            if (curr != NULL)
            {
                stack.push(curr);
                curr = curr->left;
            }
            else
            {
                curr = stack.top();
                stack.pop();
                res.push(curr->val);
                curr = curr->right;
            }
        }
    }

    int next()
    {
        // int val = res.top(); res.erase(res.begin());
        int val = res.front();
        res.pop(); // in-order traversal using queues in c++ for O(1) removal rather than O(n)
        return val;
    }

    bool hasNext()
    {
        return (res.size() > 0);
    }

private:
    // vector<int> res;
    queue<int> res;
    stack<TreeNode *> stack;
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */